pgrx/src/shmem_hash.rs

```diff
@@ -0,0 +1,250 @@
+
+        let return_value = if entry.is_null() {
+            None
+        } else {
+            let value_ptr = value_ptr!(entry);
+            let value = unsafe { std::ptr::read(value_ptr) };
+            Some(value.value)
+        };
+
+        // If we don't do this check, pg will overwrite
+        // some random entry with our key/value pair...
+        if entry.is_null() && htab.elements == self.size {
+            return Err(ShmemHashMapError::HashTableFull);
+        }
+
+        let entry = unsafe {
+            pg_sys::hash_search_with_hash_value(
+                htab.htab,
+                key_ptr,
+                hash_value,
+                pg_sys::HASHACTION_HASH_ENTER_NULL,
+                &mut found,
+            )
+        };
+
+        if !entry.is_null() {
+            let value_ptr = value_ptr!(entry);
+            let value = Value { key: Key { key }, value };
+            unsafe {
+                std::ptr::copy(std::ptr::addr_of!(value), value_ptr, 1);
+            }
+            // We inserted a new element, increasing the size of the table.
+            if return_value.is_none() {
+                htab.elements += 1;
+            }
+            Ok(return_value)
+        } else {
+            // OOM. We pre-allocate at server start, so this should never be an issue.
+            return Err(ShmemHashMapError::HashTableFull);
+        }
+    }
+
+    /// Get a value from the HashMap using the key.
+    /// If the key doesn't exist, return `None`.
+    pub fn get(&self, key: K) -> Option<V> {
+        let htab = self.htab.get().unwrap().lock();
+        let (key_ptr, hash_value) = key!(key, htab);
+
+        let entry = unsafe {
+            pg_sys::hash_search_with_hash_value(
+                htab.htab,
+                key_ptr,
+                hash_value,
+                pg_sys::HASHACTION_HASH_FIND,
+                std::ptr::null_mut(),
+            )
+        };
+
+        if entry.is_null() {
+            return None;
+        } else {
+            let value_ptr = value_ptr!(entry);
+            let value = unsafe { std::ptr::read(value_ptr) };
+            return Some(value.value);
+        }
+    }
+
+    /// Remove the value from the `ShmemHashMap` and return it.
+    /// If the key doesn't exist, return None.
+    pub fn remove(&self, key: K) -> Option<V> {
+        if let Some(value) = self.get(key) {
+            let mut htab = self.htab.get().unwrap().lock();
+            let (key_ptr, hash_value) = key!(key, htab);
+
+            // Dangling pointer, don't touch it.
+            let _ = unsafe {
+                pg_sys::hash_search_with_hash_value(
+                    htab.htab,
+                    key_ptr,
+                    hash_value,
+                    pg_sys::HASHACTION_HASH_REMOVE,
+                    std::ptr::null_mut(),
+                );
+            };
+
+            htab.elements -= 1;
+            return Some(value);
+        } else {
+            return None;
+        }
```

Comment:
But if you do use let-else this should probably be updated.

Reply:
Something like this, but then it will want you to hit it with rustfmt.
```suggestion
            Some(value)
```